File: src/rainbow/inner/coordinate/system/event/ZoomEvent.java
package rainbow.inner.coordinate.system.event;

import java.awt.event.MouseWheelEvent;

/**
 * 坐标系缩放事件
 *
 * @author Rainbow Yang
 */
public class ZoomEvent implements CoordinateSystemEvent {
    private MouseWheelEvent event;

    public ZoomEvent(MouseWheelEvent event) {
        this.event = event;
    }

    public double getZoomLevel() {
        return event.getWheelRotation();
    }

    public double getX() {
        return event.getX();
    }

    public double getY() {
        return event.getY();
    }
}

File: src/main/kotlin/rainbow/inner/coordinate/system/comp/controller/event/ZoomEvent.kt
package rainbow.inner.coordinate.system.comp.controller.event

import rainbow.inner.coordinate.point.PointDouble
import rainbow.inner.coordinate.system.CoordinateSystem
import rainbow.inner.coordinate.system.moveTo
import java.awt.event.MouseWheelEvent

/**
 * @author Rainbow Yang
 */
class ZoomEvent(event: MouseWheelEvent) : CoordinateSystemControlEvent({
    val nowLocation = it.locationChanger.toSystem(PointDouble(event.x, event.y))
    it.moveTo(nowLocation)
    it.axes.allLengthTimes *= Math.pow(zoomSpeed, event.wheelRotation.toDouble())
    it.moveTo(-nowLocation)
}
) {
    companion object {
        val zoomSpeed = 1.1
    }
}
